### DISCUSSION TITLE
Disable inline comments out of diff-scope

### DISCUSSION THREAD

**Discussion by BrunnerLivio** [danger/danger-js#1195] [Ideas]
When using e.g. `warn('Message', 'my-file.js', 203)`, Danger automatically adds an inline comment to the file if `my-file.js#203` is part of the diff in Gitlab (don't know about the other Platforms. If it is outside of the diff-scope, it will just simply add it to the PR comment.

Would it be possible to decide per message or globally whether the messages out of the "diff" should be displayed?

Some options:

*Using paramter*

```typescript
warn('Message', 'my-file.js', 203, { displayOutOfScope: false })
```

*Globally disable it*

```typescript
danger.displayOutOfScope(false)
```

*Programmatic way*

```typescript
if (danger.git.diff.includes(d => d.startLine < 203 && d.endLine > 203)) {
  warn('Message', 'my-file.js', 203)
}
```

**Comment by orta** [comment]
I don't think we need this,  you can use `warn("filename:203 something")` in that case

**Reply by BrunnerLivio** [reply]
Hmm I am sorry, I don't follow :/ 

To clarify, here is an example

In the following example I use with Danger on Gitlab

![Screenshot 2022-01-11 at 12 40 42 1](https://user-images.githubusercontent.com/9899423/148936492-a2ec02f6-aa08-4e18-bfe6-f1f3695458a2.png)


The way I understand Danger, it automatically creates inline comments for the first two statements in the code example below and
appends the last two statements to the MR comment. I assume this is because `my-file.md#L13` is part of the PR diff whereas `openapi.md` is outside of the diff. 



```typescript
// Inline comments 
warn('"Therefore" is wordy or unneeded', 'my-file.md',13)
warn('"shall" is wordy or unneeded', 'my-file.md',13)

// PR Comment
warn("'modify" is wordy or unneeded', 'openapi.md', 33)
warn("'In order to" is wordy or unneeded', 'openapi.md', 38)
```

What I want is to somehow be able **to not show the last two statements**. I only want to give warnings what folks have changed.
I am aware that Danger has the function `await danger.git.diffForFile(path)`, though the object which gets returned is really hard to work with in this case because I'd need the line numbers of the diff

**Comment by BrunnerLivio** [comment]
I was able to figure it out. It's a bit cumbersome but it works:


```typescript
async function getAddedLinesForFile(path: string): Promise<number[]> {
  const structuredDiff = await danger.git.structuredDiffForFile(path);

  const chunks = structuredDiff.chunks.reduce(
    (prev, chunk) => prev.concat(chunk.changes),
    []
  ) as { ln: number; type: string }[];

  const lines = chunks
    .filter((chunk) => chunk.type === "add")
    .map((chunk) => chunk.ln);
  return [...new Set(lines)];
}

async function printLinterMessages() {
   // The file you wanna lint
   const path = 'my-file.md';
   // Warning maybe coming from remark or another linter
   const warnings = [{ line: 13, message: '"modify" is wordy or unneeded' }]
   
  const addedLines = await getAddedLinesForFile(path);
  
  warnings
    // Only get the warnings which were added by the user
    .filter(w => addedLines.include(w.line))
    .forEach(w => warn(w.message, path, w.line)) 
}
